[Viewport Clipping] [Part 3/4] Plumb the viewport anchor layer through the scrolling tree

whsieh opened the pull request:
#### 7de5d103812b0f8da9ff696e9058bc13011a6da3
<pre>
[Viewport Clipping] [Part 3/4] Plumb the viewport anchor layer through the scrolling tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=292124">https://bugs.webkit.org/show_bug.cgi?id=292124</a>
<a href="https://rdar.apple.com/150150239">rdar://150150239</a>

Reviewed by Abrar Rahman Protyasha.

Continue working towards clipping fixed and sticky elements to the viewport. In this patch, we add
plumbing for the viewport anchor layer through scrolling tree and scrolling state nodes. At the
moment, this viewport anchor layer is redundant with the scrolling tree/state node&apos;s layer; however,
in the next and final patch, we&apos;ll add a new clipping layer which will replace the anchor layer as
the scrolling node&apos;s primary layer, in the case where the fixed or sticky element is clipped to the
viewport.

See below for more details; no change in behavior yet.

* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setNodeLayers):

Add `viewportAnchorLayer` to `NodeLayers`.

* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.h:

Add a changed property flag corresponding to the viewport anchor layer.

* Source/WebCore/page/scrolling/ScrollingStateStickyNode.cpp:
(WebCore::ScrollingStateStickyNode::ScrollingStateStickyNode):
(WebCore::ScrollingStateStickyNode::applicableProperties const):
(WebCore::ScrollingStateStickyNode::setViewportAnchorLayer):
(WebCore::ScrollingStateStickyNode::computeAnchorLayerPosition const):

Split `computeLayerPosition` into two methods: one to compute the position of the anchor layer, and
another to compute the position of the clipping layer (to be introduced in the next patch). Note
that both of these positions are relative to the sticky node&apos;s parent layer.

(WebCore::ScrollingStateStickyNode::computeClippingLayerPosition const):

Add a new method to return the position of the (soon-to-exist) clipping layer.

(WebCore::ScrollingStateStickyNode::reconcileLayerPositionForViewportRect):

Teach this to update both the viewport anchor layer as well as the viewport clipping layer. In the
case where the clipping layer doesn&apos;t exist, we fall back to existing behavior (i.e. blitting the
anchor layer to the anchor layer position); however, if the clipping layer exists, we&apos;ll instead
blit the clipping layer and then counter-blit the anchor layer as needed.

(WebCore::ScrollingStateStickyNode::hasViewportClippingLayer const):

Add a helper method to return whether or not the scrolling node has a viewport clipping layer (i.e.
whether the primary scrolling node layer is different from the viewport anchor layer).

(WebCore::ScrollingStateStickyNode::scrollDeltaSinceLastCommit const):

Use the clipping layer here when computing scroll delta, if it exists.

(WebCore::ScrollingStateStickyNode::computeLayerPosition const): Deleted.
* Source/WebCore/page/scrolling/ScrollingStateStickyNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeStickyNode.cpp:
(WebCore::ScrollingTreeStickyNode::computeClippingLayerPosition const):

Add a new method to return the position of the (soon-to-exist) clipping layer, by calling into
`ScrollingStateViewportConstrainedNode::computeLayerPosition`.

(WebCore::ScrollingTreeStickyNode::scrollDeltaSinceLastCommit const):

Match `ScrollingStateStickyNode::scrollDeltaSinceLastCommit`.

* Source/WebCore/page/scrolling/ScrollingTreeStickyNode.h:
(WebCore::ScrollingTreeStickyNode::hasViewportClippingLayer const):
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.mm:
(WebCore::ScrollingTreeStickyNodeCocoa::commitStateBeforeChildren):

Set `m_viewportAnchorLayer` based on the sticky state node&apos;s viewport anchor layer representation.

(WebCore::ScrollingTreeStickyNodeCocoa::applyLayerPositions):

Match the changes in `reconcileLayerPositionForViewportRect` above, by only blitting the anchor
layer to its updated position in the case where the clipping layer doesn&apos;t exist; otherwise, blit
the containing clipping layer to match the viewport rect while scrolling, and simultaneously
counter-blit the anchor layer underneath.

(WebCore::ScrollingTreeStickyNodeCocoa::hasViewportClippingLayer const):

Add a helper method to return whether or not the scrolling node has a viewport clipping layer (i.e.
whether the primary scrolling node layer is different from the viewport anchor layer).

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateDebugIndicators):

Drive-by fix: fix a typo in this comment.

(WebCore::RenderLayerBacking::childForSuperlayersExcludingViewTransitions const):

Use the more generic `viewportClippingOrAnchorLayer()` here, instead of returning the viewport
anchor layer. This method just returns `m_viewportAnchorLayer` right now, but once the clipping
layer is introduced, it will return the clipping layer instead (which will contain the anchor
layer).

* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::didChangePlatformLayerForLayer):

Set both the new `viewportAnchorLayer` alongside `layer` on `NodeLayers`; these are always identical
for now, but in the next patch will change such that `layer` points to the outer layer that needs to
be repositioned when the scrolling node changes, while `viewportAnchorLayer` will continue to point
to just the anchor layer.

(WebCore::RenderLayerCompositor::computeFixedViewportConstraints const):
(WebCore::RenderLayerCompositor::computeStickyViewportConstraints const):

Adopt `viewportClippingOrAnchorLayer` here, instead of just using `viewportAnchorLayer` to compute
rects for the viewport constraints.

(WebCore::RenderLayerCompositor::updateScrollingNodeForViewportConstrainedRole):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::connectStateNodeLayers):

Add Cocoa-specific logic to update scrolling state tree nodes with the viewport anchor layer, by
mapping layer ID to platform `CALayer`.

Canonical link: <a href="https://commits.webkit.org/294311@main">https://commits.webkit.org/294311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b641f3d513aa47aa4032dd39a2443d048166e1a3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101388 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21053 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11358 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106541 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52017 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103428 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21360 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29546 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77220 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34260 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104395 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16486 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91577 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57562 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16307 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9589 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51365 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/86186 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9663 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108893 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28517 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20984 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86195 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28879 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87779 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85754 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21818 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30487 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8196 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22634 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28448 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33727 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28259 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31579 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29818 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/scrolling/ScrollingStateStickyNode.cpp:
```diff
@@ -69,13 +72,25 @@ Ref<ScrollingStateNode> ScrollingStateStickyNode::clone(ScrollingStateTree& adop
 OptionSet<ScrollingStateNode::Property> ScrollingStateStickyNode::applicableProperties() const
 {
-    constexpr OptionSet<Property> nodeProperties = { Property::ViewportConstraints };
+    constexpr OptionSet<Property> nodeProperties = {
```
Nit:
```suggestion
    constexpr OptionSet nodeProperties {
```

whsieh replied:
(Will also make it `static constexpr`)

aprotyas commented on Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.mm:
```diff
@@ -53,26 +53,43 @@
+    if (hasViewportClippingLayer()) {
+        auto clippingLayerPosition = computeClippingLayerPosition();
+        [m_layer _web_setLayerTopLeftPosition:clippingLayerPosition];
+        LOG_WITH_STREAM(Scrolling, stream << "ScrollingTreeStickyNodeCocoa " << scrollingNodeID() << " constrainingRectAtLastLayout " << m_constraints.constrainingRectAtLastLayout() << " last layer pos " << m_constraints.layerPositionAtLastLayout() << " clippingLayerPosition " << clippingLayerPosition);
```
Nit: it may help to reference the function we're logging this from?
```suggestion
        LOG_WITH_STREAM(Scrolling, stream << "ScrollingTreeStickyNodeCocoa::applyLayerPositions() " << scrollingNodeID() << " constrainingRectAtLastLayout " << m_constraints.constrainingRectAtLastLayout() << " last layer pos " << m_constraints.layerPositionAtLastLayout() << " clippingLayerPosition " << clippingLayerPosition);
```

whsieh replied:
(I think I'll also generally clean this up, so that we log the clipping/anchor layer positions separately — that would help avoid redundant logs for the first part of the log stream)

aprotyas commented on Source/WebCore/rendering/RenderLayerBacking.cpp:
```diff
@@ -3543,8 +3543,8 @@ GraphicsLayer* RenderLayerBacking::childForSuperlayersExcludingViewTransitions()
-    if (m_viewportAnchorLayer)
-        return m_viewportAnchorLayer.get();
+    if (auto* viewportConstrainedLayer = viewportClippingOrAnchorLayer())
+        return viewportConstrainedLayer;
```
Maybe?
```suggestion
    if (RefPtr viewportConstrainedLayer = viewportClippingOrAnchorLayer())
        return viewportConstrainedLayer.get();
```

whsieh replied:
Oh, sure — I can adopt smart pointers here.

(The only reason I avoided it was because `viewportClippingOrAnchorLayer()` is an inline function, so it's technically not necessary for the checker)

whsieh committed:
[Viewport Clipping] [Part 3/4] Plumb the viewport anchor layer through the scrolling tree
https://bugs.webkit.org/show_bug.cgi?id=292124
rdar://150150239

Reviewed by Abrar Rahman Protyasha.

Continue working towards clipping fixed and sticky elements to the viewport. In this patch, we add
plumbing for the viewport anchor layer through scrolling tree and scrolling state nodes. At the
moment, this viewport anchor layer is redundant with the scrolling tree/state node's layer; however,
in the next and final patch, we'll add a new clipping layer which will replace the anchor layer as
the scrolling node's primary layer, in the case where the fixed or sticky element is clipped to the
viewport.

See below for more details; no change in behavior yet.

* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setNodeLayers):

Add `viewportAnchorLayer` to `NodeLayers`.

* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.h:

Add a changed property flag corresponding to the viewport anchor layer.

* Source/WebCore/page/scrolling/ScrollingStateStickyNode.cpp:
(WebCore::ScrollingStateStickyNode::ScrollingStateStickyNode):
(WebCore::ScrollingStateStickyNode::applicableProperties const):
(WebCore::ScrollingStateStickyNode::setViewportAnchorLayer):
(WebCore::ScrollingStateStickyNode::computeAnchorLayerPosition const):

Split `computeLayerPosition` into two methods: one to compute the position of the anchor layer, and
another to compute the position of the clipping layer (to be introduced in the next patch). Note
that both of these positions are relative to the sticky node's parent layer.

(WebCore::ScrollingStateStickyNode::computeClippingLayerPosition const):

Add a new method to return the position of the (soon-to-exist) clipping layer.

(WebCore::ScrollingStateStickyNode::reconcileLayerPositionForViewportRect):

Teach this to update both the viewport anchor layer as well as the viewport clipping layer. In the
case where the clipping layer doesn't exist, we fall back to existing behavior (i.e. blitting the
anchor layer to the anchor layer position); however, if the clipping layer exists, we'll instead
blit the clipping layer and then counter-blit the anchor layer as needed.

(WebCore::ScrollingStateStickyNode::hasViewportClippingLayer const):

Add a helper method to return whether or not the scrolling node has a viewport clipping layer (i.e.
whether the primary scrolling node layer is different from the viewport anchor layer).

(WebCore::ScrollingStateStickyNode::scrollDeltaSinceLastCommit const):

Use the clipping layer here when computing scroll delta, if it exists.

(WebCore::ScrollingStateStickyNode::computeLayerPosition const): Deleted.
* Source/WebCore/page/scrolling/ScrollingStateStickyNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeStickyNode.cpp:
(WebCore::ScrollingTreeStickyNode::computeClippingLayerPosition const):

Add a new method to return the position of the (soon-to-exist) clipping layer, by calling into
`ScrollingStateViewportConstrainedNode::computeLayerPosition`.

(WebCore::ScrollingTreeStickyNode::scrollDeltaSinceLastCommit const):

Match `ScrollingStateStickyNode::scrollDeltaSinceLastCommit`.

* Source/WebCore/page/scrolling/ScrollingTreeStickyNode.h:
(WebCore::ScrollingTreeStickyNode::hasViewportClippingLayer const):
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.h:
* Source/WebCore/page/scrolling/cocoa/ScrollingTreeStickyNodeCocoa.mm:
(WebCore::ScrollingTreeStickyNodeCocoa::commitStateBeforeChildren):

Set `m_viewportAnchorLayer` based on the sticky state node's viewport anchor layer representation.

(WebCore::ScrollingTreeStickyNodeCocoa::applyLayerPositions):

Match the changes in `reconcileLayerPositionForViewportRect` above, by only blitting the anchor
layer to its updated position in the case where the clipping layer doesn't exist; otherwise, blit
the containing clipping layer to match the viewport rect while scrolling, and simultaneously
counter-blit the anchor layer underneath.

(WebCore::ScrollingTreeStickyNodeCocoa::hasViewportClippingLayer const):

Add a helper method to return whether or not the scrolling node has a viewport clipping layer (i.e.
whether the primary scrolling node layer is different from the viewport anchor layer).

* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateDebugIndicators):

Drive-by fix: fix a typo in this comment.

(WebCore::RenderLayerBacking::childForSuperlayersExcludingViewTransitions const):

Use the more generic `viewportClippingOrAnchorLayer()` here, instead of returning the viewport
anchor layer. This method just returns `m_viewportAnchorLayer` right now, but once the clipping
layer is introduced, it will return the clipping layer instead (which will contain the anchor
layer).

* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::didChangePlatformLayerForLayer):

Set both the new `viewportAnchorLayer` alongside `layer` on `NodeLayers`; these are always identical
for now, but in the next patch will change such that `layer` points to the outer layer that needs to
be repositioned when the scrolling node changes, while `viewportAnchorLayer` will continue to point
to just the anchor layer.

(WebCore::RenderLayerCompositor::computeFixedViewportConstraints const):
(WebCore::RenderLayerCompositor::computeStickyViewportConstraints const):

Adopt `viewportClippingOrAnchorLayer` here, instead of just using `viewportAnchorLayer` to compute
rects for the viewport constraints.

(WebCore::RenderLayerCompositor::updateScrollingNodeForViewportConstrainedRole):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::connectStateNodeLayers):

Add Cocoa-specific logic to update scrolling state tree nodes with the viewport anchor layer, by
mapping layer ID to platform `CALayer`.

Canonical link: https://commits.webkit.org/294311@main
